confirms.go

```diff
@@ -101,3 +109,56 @@ func (c *confirms) Close() error {
 	c.listeners = nil
 	return nil
 }
+
+type deferredConfirmations struct {
+	m             sync.Mutex
+	confirmations map[uint64]*DeferredConfirmation
+}
+
+func newDeferredConfirmations() *deferredConfirmations {
+	return &deferredConfirmations{
+		confirmations: map[uint64]*DeferredConfirmation{},
+	}
+}
+
+func (d *deferredConfirmations) Add(tag uint64) *DeferredConfirmation {
+	d.m.Lock()
+	defer d.m.Unlock()
+
+	dc := &DeferredConfirmation{DeliveryTag: tag}
+	dc.wg.Add(1)
+	d.confirmations[tag] = dc
+	return dc
+}
+
+func (d *deferredConfirmations) Confirm(confirmation Confirmation) {
+	d.m.Lock()
+	defer d.m.Unlock()
+
+	dc, found := d.confirmations[confirmation.DeliveryTag]
+	if !found {
+		// we should never receive a confirmation for a tag that hasn't been published, but a test causes this to happen
+		return
+	}
+	dc.confirmation = confirmation
+	dc.wg.Done()
+	delete(d.confirmations, confirmation.DeliveryTag)
+}
+
+func (d *deferredConfirmations) ConfirmMultiple(confirmation Confirmation) {
+	d.m.Lock()
+	defer d.m.Unlock()
+
+	for k, v := range d.confirmations {
```

Comment:
A potential optimization I decided against implementing was to keep track of which confirmations are outstanding, which would allow us to do a O(1) lookup on this map rather than looping through every key. My hunch is that this would not be that much of a bottleneck, but it's pretty unfounded.

Reply:
we scanning here over whole d.confirmations because range on maps is unordered, can it be a potentiall hot spot?

Reply:
Yeah, if we were to have a huge amount of pending confirmations and received lots of small multi-confirms this could be inefficient, since we would be cycling through the whole map for each multi-confirm but skipping most of the keys.

My hunch though is that:
1. If there's a large amount of pending confirmations Rabbit will probably send a small number of large-range multi-confirms
2. We generally won't have a large amount of pending confirmations